openerp/openupgrade/openupgrade.py

```diff
@@ -622,3 +625,167 @@ def wrapped_function(cr, version):
+
+            # All other records are to remove
+            old_record_ids = [z.id for z in grouped_records[1:]]
+
+            all_record_ids = old_record_ids + [new_record_id]
+
+            # Replace every many2one record in the database that has an old
+            # record as value with the record to keep
+
+            for table in tables_to_lookup:
+                table_name = table[0]
+
+                # Prevent the upgrade script to create duplicates
+                # in the many2many relation table and raise a constraint error
+                if table[2] == 'many2many':
+                    cr.execute(
+                        "   SELECT t.%(other_field)s, t.%(field_name)s "
+                        "   FROM %(table_name)s as t"
+                        "   WHERE %(field_name)s "
+                        "   in %(all_record_ids)s "
+                        "   ORDER BY %(other_field)s" %
+                        {
+                            'table_name': table_name,
+                            'field_name': table[1],
+                            'other_field': table[3],
+                            'all_record_ids': tuple(all_record_ids),
+                        })
+
+                    for k, group_to_check in groupby(
+                        cr.fetchall(), lambda rec: rec[0]
+                    ):
+                        group_to_check = list(group_to_check)
+                        if len(group_to_check) > 1:
+                            for rec_to_unlink in group_to_check[1:]:
+                                cr.execute(
+                                    "   DELETE FROM %(table_name)s "
+                                    "   WHERE %(field_name)s = %(field_value)s"
+                                    "   AND %(other_field)s "
+                                    "   = %(other_field_value)s" %
+                                    {
+                                        'table_name': table_name,
+                                        'field_name': table[1],
+                                        'field_value': rec_to_unlink[1],
+                                        'other_field': table[3],
+                                        'other_field_value': rec_to_unlink[0],
+                                    })
+
+                # Main upgrade script
+                cr.execute(
+                    "   UPDATE %(table_name)s"
+                    "   SET %(field_name)s = %(new_value)s"
+                    "   WHERE %(field_name)s %(old_record_ids)s;" %
+                    {
+                        'table_name': table_name,
+                        'field_name': table[1],
+                        'new_value': new_record_id,
+                        'old_record_ids': len(old_record_ids) > 1 and
+                        'in %s' % (tuple(old_record_ids),) or '= %s' %
+                        old_record_ids[0]
+                    })
+
+            model_pool.unlink(cr, SUPERUSER_ID, old_record_ids)
```

Comment:
Please log info about the unlinked records to track any information loss that might occur.